### PR TITLE
USHIFT-1312 Test: Rebooting healthy system should result in backing up the data

### PR DESCRIPTION
**Pull request by pmtk**
Tests functionality of https://github.com/openshift/microshift/pull/1874

### CODE REVIEW

**Comment by pmtk** [comment]
This test will probably require setting up current `microshift-e2e` to ignore `ostree` tags or provide mechanism to not run tests on non-ostree-based systems

File: test/suites/upgrade_ostree.robot [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+    [Documentation]    Checks if property of a service has expected value
+    [Arguments]    ${service}    ${property}    ${expected_value}
+
+    ${value}    ${rc}=    Execute Command
+    ...    systemctl show ${service} -P ${property}
+    ...    sudo=True    return_rc=True
+    Should Be Equal As Integers    0    ${rc}
+    Should Be Equal As Strings    ${expected_value}    ${value}
+
+Reboot
```
Comment (pmtk) [inline test/suites/upgrade_ostree.robot:67]: From @dhellmann ([src](https://github.com/openshift/microshift/pull/1826#discussion_r1212161152))
> This could go into the microshift-host resource file.

File: test/suites/upgrade_ostree.robot [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+    [Documentation]    Test suite teardown
+    Logout MicroShift Host
+
+Wait Until Greenboot Health Checks Are Done
+    [Documentation]    Wait until greenboot healthchecks are done
+
+    Wait Until Keyword Succeeds    5m    15s
+    ...    Check Service Property    greenboot-healthcheck.service    SubState    exited
+
+Wait Until MicroShift Is Ready
```
Comment (pmtk) [inline test/suites/upgrade_ostree.robot:51]: From @dhellmann ([src](https://github.com/openshift/microshift/pull/1826#discussion_r1212164372)) 
> There's a similar keyword in my PR with etcd tests that does a bit more. https://github.com/openshift/microshift/pull/1735/files#diff-4f6b7fe814ca7246b4cc856e49319e6e2f33ba6ac2abe86995e7a42cafab7afcR36
>
> I'd like to make that reusable keyword as robust as we can.

File: test/suites/upgrade_ostree.robot [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+    [Documentation]    Wait until MicroShift service is ready
+
+    Wait Until Keyword Succeeds    2m    15s
+    ...    Check Service Property    microshift.service    ActiveState    active
+
+Check Service Property
```
Comment (pmtk) [inline test/suites/upgrade_ostree.robot:57]: From @dhellmann ([src](https://github.com/openshift/microshift/pull/1826#discussion_r1212165319)):
> See https://github.com/openshift/microshift/pull/1735/files#diff-9f059e74d79eab4cda8c994b93013638f88857b968777109b9bc462d89fa5f7fR10 as well

**Comment by pmtk** [comment]
/cc @dhellmann @pacevedom @jogeo @dhensel-rh 

This PR needs some sort mechanism to not run tests intended for R4E/ostree-based systems on RPM.
Approaches that come to mind:
- doing a runtime check and deciding if test should run or be skipped
- using tags to filter out irrelevant test suites
any other ideas?

File: test/resources/microshift-host.resource [hunk 4]
```diff
@@ -25,3 +25,15 @@ Login MicroShift Host
 Logout MicroShift Host
     [Documentation]  Close the open ssh connection to the MicroShift host
     SSHLibrary.Close Connection
+
+Reboot MicroShift Host
```
Comment (dhellmann) [inline test/resources/microshift-host.resource:29]: This would be more robust if we looked for a different boot id after the reboot, to avoid any race condition with issuing the reboot command and sshd stopping. It doesn't seem like the most likely race, so let's make that change in another PR.

File: test/suites/upgrade-ostree.robot [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+Remove Existing Backup For Current Deployment
+    [Documentation]    If exists, remove backup for current deployment
+
+    ${exists}=    Does Backup For Booted Deployment Exist
+    IF    ${exists} == ${False}    RETURN
+
+    ${deployID}=    Get Booted Deployment ID
+
+    ${rm_output}    ${rm_stderr}    ${rc}=    Execute Command
+    ...    find ${BACKUP_STORAGE} -name '${deployID}*' -exec rm -rf {} +
```
Comment (dhellmann) [inline test/suites/upgrade-ostree.robot:71]: We know the exact path, right? Can we just remove it, instead of using `find`?

**Review by dhellmann — COMMENTED** [review]
This looks OK, but I see the e2e test is failing with an error in this new test suite

```
Keyword 'Greenboot Health Check Exited' failed after retrying for 5 minutes. The last error was: dead != exited
```

File: test/suites/upgrade-ostree.robot (hunk 5 above)
Comment (dhellmann) [reply test/suites/upgrade-ostree.robot:71]: OK, we probably want to do something to limit the depth of the find call, then. Is it enough to say `rm -rf ${deployID}*` instead of using find at all?

And we probably want to check for deployID being empty, to ensure we don't delete everything.

Comment (pmtk) [reply test/suites/upgrade-ostree.robot:71]: `find` was only looking into `/var/lib/microshift-backups/` but I changed it to `rm -rf` because it's easier to read

File: test/suites-ostree/upgrade-ostree.robot [hunk 6]
```diff
@@ -0,0 +1,82 @@
+*** Settings ***
```
Comment (pmtk) [inline test/suites-ostree/upgrade-ostree.robot:1]: This file is in separate dir so e2e CI job doesn't try to run it against non-ostree system.
When we have e2e for ostree systems we can think if we want to move it back to `suites/` and use tags (or other method to filter tests)

**Comment by pmtk** [comment]
/remove-label tide/merge-method-squash

Remembered that tide's squash results in ugly commit messages. I'll squash the commits myself

**Review by dhellmann — COMMENTED** [review]
This looks good to me. I'll leave it for @jogeo to approve since he wanted a chance to review, too.

File: test/resources/ostree.resource [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+    Should Not Be Empty    ${current_deployment_id}
+    Should Not Be Empty    ${BACKUP_STORAGE}
+    ${backup_path}=    Catenate    SEPARATOR=/    ${BACKUP_STORAGE}    ${current_deployment_id}
+    RETURN    ${backup_path}
+
+Get System Health
+    [Documentation]    Get system health information from health.json file
+
+    ${health}    ${stderr}    ${rc}=    Execute Command
+    ...    jq -r '.health' ${BACKUP_STORAGE}/health.json
```
Comment (jogeo) [inline test/resources/ostree.resource:46]: 'Evaluate    json.loads' was used above in keyword "Get Booted Deployment Status". Should we be consistent in the pattern we use to ingest and parse json?

File: test/suites-ostree/upgrade-ostree.robot [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+    [Documentation]    Checks if greenboot-healthcheck finished running
+
+    ${value}=    Get Systemd Setting    greenboot-healthcheck.service    SubState
+    Should Be Equal As Strings    ${value}    exited
+
+Remove Existing Backup For Current Deployment
+    [Documentation]    Remove backups for current deployment
+
+    ${path}=    Get Booted Deployment Backup Prefix Path
+
```
Comment (jogeo) [inline test/suites-ostree/upgrade-ostree.robot:61]: Should we test path here with 'Should Not Be Empty'? I see that 'Get Booted Deployment Backup Prefix Path' has a check for empty values. However, if the code is changed to use a different key word at some point to get path we would not want to accidentally call 'rm -rf *'

**Review by jogeo — COMMENTED** [review]
Thank you for adding this test, it looks good over all. I've left a few questions and comments.

File: test/resources/ostree.resource (hunk 7 above)
Comment (jogeo) [reply test/resources/ostree.resource:46]: As a side note, 'jq' is used by scripts such as microshift_set_healthy.sh, so it's not a bad thing that we are also using it in our test scripts. My initial comment was just a reaction to seeing json parsed in the same PR using two different approaches.

File: test/resources/ostree.resource [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+
+*** Keywords ***
+Get Booted Deployment Status
+    [Documentation]    Get rpm-ostree status of booted deployment as a json
+
+    ${ostree_status}    ${rc}=    Execute Command
+    ...    rpm-ostree status --booted --json
+    ...    return_rc=True
+    Should Be Equal As Integers    0    ${rc}
+    ${json}=    Evaluate    json.loads('''${ostree_status}''')    json
```
Comment (dhellmann) [inline test/resources/ostree.resource:20]: For YAML, I found it handy to parse the text and then return a RF DottedDict type to make accessing the data from RF keywords easier. See the YAML library in the resources directory.

File: test/resources/ostree.resource [hunk 10]
```diff
@@ -0,0 +1,62 @@
...
+    RETURN    ${json}
+
+Get Booted Deployment ID
+    [Documentation]    Get ID of currently booted deployment
+
+    ${status}=    Get Booted Deployment Status
+    RETURN    ${status["deployments"][0]["id"]}
```
[the remark this answers is not in the document]
Comment (pmtk) [reply test/resources/ostree.resource:27]: This uses kw `Get Booted Deployment Status` which runs `rpm-ostree status --booted --json`
Thanks to `--booted` it will only output one deployment, but it's still using the same data structure, i.e. `deployments` in an array. With `--booted` it will only have 1 element.

Without that switch, `0` could be either booted or staged deployment

File: test/resources/ostree.resource (hunk 7 above)
Comment (pmtk) [reply test/resources/ostree.resource:46]: That's a good point. `jq` here felt simpler, otherwise I'd have either `cat` and use `json.loads` or... I don't know what, I don't think `SSHLibrary.Get File` supports privilege escalation and `health.json` is only readable by root

Comment (dhellmann) [reply test/resources/ostree.resource:47]: I like the idea of keywords for grabbing files only readable by root. [Save Default MicroShift Config](https://github.com/openshift/microshift/blob/main/test/resources/microshift-config.resource#L13) uses `cat` to grab the contents of the config file, and it could use the same reusable keyword for that step. Then something like [Upload String To File](https://github.com/openshift/microshift/blob/main/test/resources/microshift-config.resource#L13) could be modified to take arguments for the owner and mode of the output file.

**Comment by ggiguash** [comment]
Let's merge this in the current form. The tests are not going to run because they are in a separate folder.
We now have a metal environment to run this tests, so let's try those out.
/lgtm
